### PR TITLE
Advanced search

### PR DESCRIPTION
**Pull request by MadameSheema**
Solves: https://github.com/franksparks/BookJourney/issues/11

Screenshot 2024-08-04 at 17 03 42

## Summary

In this PR we are implementing the advanced search:

- When a user clicks on `All results` from the preview search is redirected to the advanced search page
- The advanced search page is displayed returning the first 10 results from the preview search
- From the advanced search page a new search by title or author can be performed

What it has not been developed yet:
- The actions on the `Want to read` button

## Implementation

* The header has been moved to a layout, in that way is present everywhere.
* From the preview search we redirect to the advanced search page

* On the advanced search page we are using different states:
  * results: we store the response from the API to display the books
  * query: the original query that is coming from the preview searchbox
  * page: the current page where we are placed for pagination purposes
  * totaltems: how many books are available, used for the pagination
  * advancedQuery: the new query we use from the advanced query page
  * radioValue: the radio value selected use to know which advanced search it should be performed
  * avoidSearch: used to avoid to perform a regular search in some specific scenarios 
  * avoidAdvancedSearch: used to avoid to perform an advanced search in some specific scenarios

Each time that the page is loaded or the query of the URL changes, we get the query from the URL, we make sure that we don't avoid any search to happen and we setup the pagination to it original state, storing the query from the URL, this will force the search to happen and the results to be displayed.
  
For an advanced query to take place, a new value should be input on the searchbox, from there and taking into consideration which is the value selected on the radio buttons, the advanced search will take place.
  
## To improve

- As it has been developed using flex, the width varies, would be great to decide how it should really look like.
- The look & feel of the dropdown since we are using shadcn and is affected by the global CSS breaking the original style.

### CODE REVIEW

File: components/SearchBox.tsx [hunk 1]
```diff
@@ -1,31 +1,65 @@
...
+    setAdvancedQuery: (advancedQuery: string) => void,
+    handleAdvancedSearch: () => void,
+    setPage: (page: number) => void,
+    setTotalItems: (items: number) => void,
+    setAvoidSearch: (status: boolean) => void,
+    setAvoidAdvancedSearch: (status: boolean) => void,
+    setRadioValue: (option: string) => void,
 }
 
 // Todo: Change the img for the next.js Image component
```
Comment (vedderzeznick) [inline components/SearchBox.tsx:21]: clean this comment

File: components/SearchResults.tsx [hunk 2]
```diff
@@ -9,18 +10,16 @@ type SearchResultProps = {
 // Todo: Change the img for the next.js Image component
```
Comment (vedderzeznick) [inline components/SearchResults.tsx:10]: clean this comment

File: app/search/page.tsx [hunk 3]
```diff
@@ -0,0 +1,105 @@
+"use client"
+
+import { Book, actionSearchBooks } from "@/actions/search-books";
```
Comment (vedderzeznick) [inline app/search/page.tsx:3]: we will have to move the "Book" model to a different folder, its not looking good if we export it from "actions", maybe something like "extra models"

File: components/PreviewSearchBox.tsx [hunk 4]
```diff
@@ -52,36 +55,44 @@ export default function PreviewSearchBox() {
...
-              <img
-                src="../default_cover.jpg"
-                alt={`Unknown cover for: ${option.label}`}
-                style={{ width: 50, height: 75, marginRight: 10, objectFit: 'cover' }}
-              />
-            )}
+            <img
+              src={option.imageUrl || "../default_cover.jpg"}
+              alt={option.label}
+              style={{ width: 50, height: 75, marginRight: 10, objectFit: 'cover' }}
```
Comment (vedderzeznick) [inline components/PreviewSearchBox.tsx:88]: we need to use classes instead of inline styles

File: components/SearchRadioButtons.tsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import FormControlLabel from '@mui/material/FormControlLabel';
+import FormControl from '@mui/material/FormControl';
+
+type searchRadioButtonsProps = {
+  handleRadioButtonChange: (event: React.SyntheticEvent) => void,
+}
+
+export default function SearchRadioButtons({ handleRadioButtonChange }: searchRadioButtonsProps) {
+
+  const radioStyles = {
```
Comment (vedderzeznick) [inline components/SearchRadioButtons.tsx:14]: we need to move these styles to classes

File: components/SearchResults.tsx [hunk 6]
```diff
@@ -9,18 +10,16 @@ type SearchResultProps = {
...
-                    <TableRow key={index}>
-                        <>{result.smallThumbnail && <TableCell><img src={result.smallThumbnail} /></TableCell>}
-                        </>
-                        <TableCell>{capitalizeFirstLetter(result.title)}</TableCell>
-                        <TableCell>
-                            {result.authors && (
-                                capitalizeFirstLetter(result.authors.join(" "))
-                            )}</TableCell>
+                    <TableRow className="items-center w-3/4" key={index}>
+                        <TableCell><img src={result.smallThumbnail || "../default_cover.jpg"} alt={`{result.title}cover`} style={{ width: '125px', height: '200px' }} /></TableCell>
```
Comment (vedderzeznick) [inline components/SearchResults.tsx:17]: avoid using inline styles

File: components/header.tsx [hunk 7]
```diff
@@ -9,32 +12,37 @@ import {
...
       <div className="mx-12 flex items-center justify-between h-full">
         <div className="flex items-center gap-x-3">
-          <div className="flex items-center gap-x-1 cursor-default">
+          <div onClick={() => router.push('/')} className="flex items-center gap-x-1 cursor-pointer">
             <h1 className="text-white text-3xl font-thin">book</h1>
             <h1 className="text-orange-500 text-3xl tracking-wide">journey</h1>
           </div>
         </div>
         <PreviewSearchBox />
         <div style={{ minWidth: '150px' }}>
```
Comment (vedderzeznick) [inline components/header.tsx:30]: we need to use classes instead of inline styles

File: package.json [hunk 8]
```diff
@@ -14,6 +14,7 @@
     "@emotion/styled": "^11.11.5",
```
Comment (vedderzeznick) [inline package.json:14]: Im seeing we're using @emotion and also @mui , I think we can make it only with MUI, but anyways, we need to separate classes in different folders and files

File: components/SearchBox.tsx (hunk 1 above)
Comment (MadameSheema) [reply components/SearchBox.tsx:21]: This comment is here to don't forget to do a task in a near future.

**Review by franksparks — APPROVED** [review]
LGTM - will try to refactor the pages in the future

**Pull request merged**